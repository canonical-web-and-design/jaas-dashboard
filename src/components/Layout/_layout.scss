@import "vanilla-framework/scss/settings_colors";

.l-container {
  display: grid;
  min-height: calc(100vh - 48px);
}

.l-container--sidebar {
  @extend .l-container;

  grid-template-columns: 150px 1fr;
}

.l-side {
  background: $color-light;
  padding: 0.5rem;
  padding-left: 0;
}

.main-content {
  display: grid;
}
