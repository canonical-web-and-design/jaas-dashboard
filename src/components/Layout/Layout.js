import React, { useEffect, useState, useRef } from "react";
import { useSelector } from "react-redux";
import Notification from "@canonical/react-components/dist/components/Notification/Notification";
import PrimaryNav from "components/PrimaryNav/PrimaryNav";
import classNames from "classnames";
import useHover from "hooks/useHover";
import { getViewportWidth, debounce } from "app/utils";

import { isSidebarCollapsible } from "ui/selectors";

import "./_layout.scss";

const Layout = ({ children }) => {
  const [sidebarRef, isSidebarHovered] = useHover();
  const [screenWidth, setScreenWidth] = useState(getViewportWidth());
  const sidebarCollapsible = useSelector(isSidebarCollapsible);
  const smallScreenBreakpoint = 1400;
  const isSmallScreen = screenWidth <= smallScreenBreakpoint ? true : false;
  let isMounted = useRef(false);

  const handleScreenResize = () => {
    if (isMounted.current) {
      setScreenWidth(getViewportWidth());
    }
  };

  useEffect(() => {
    isMounted.current = true;
    window.addEventListener("resize", debounce(handleScreenResize, 250));
    return () => {
      isMounted.current = false;
      window.removeEventListener("resize", debounce(handleScreenResize, 250));
    };
  }, []);

  return (
    <div
      className={classNames("l-container", {
        "has-collapsible-sidebar": sidebarCollapsible || isSmallScreen,
      })}
    >
      <div
        className={classNames("l-side", {
          "is-collapsed":
            (sidebarCollapsible && !isSidebarHovered) ||
            (isSmallScreen && !isSidebarHovered),
        })}
        ref={sidebarRef}
      >
        <PrimaryNav />
      </div>
      <main className="l-main" id="main-content">
        <div data-test="main-children">{children}</div>
        {!localStorage.getItem("release20_04") && (
          <Notification
            type="information"
            close={() => {
              localStorage.setItem("release20_04", true);
            }}
          >
            Welcome the new JAAS Dashboard! This dashboard is the replacement
            for the Juju GUI in JAAS and individual Juju Controllers from Juju
            2.8.{" "}
            <span className="u-hide--small">
              Read more and join the discussion about this new product{" "}
              <a href="#_">on Discourse</a>. We would love to hear your
              feedback!
            </span>
          </Notification>
        )}
      </main>
    </div>
  );
};

export default Layout;
