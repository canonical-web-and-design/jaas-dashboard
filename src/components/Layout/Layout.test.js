import React from "react";
import { shallow } from "enzyme";

import Layout from "./Layout";

describe("Layout", () => {
  it("renders without crashing and matches snapshot", () => {
    const wrapper = shallow(<Layout />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders with a sidebar", () => {
    const wrapper = shallow(<Layout />);
    expect(wrapper.find(".l-side")).toHaveLength(1);
  });

  it("renders without a sidebar if false bool is passed", () => {
    const wrapper = shallow(<Layout sidebar={false} />);
    expect(wrapper.find(".l-side")).toHaveLength(0);
  });

  it("should display the children", () => {
    const wrapper = shallow(<Layout>content</Layout>);
    expect(wrapper.find("#main-content").html()).toStrictEqual(
      `<main class="main-content" id="main-content">content</main>`
    );
  });
});
