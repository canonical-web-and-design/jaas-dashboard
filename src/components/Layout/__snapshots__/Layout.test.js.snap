// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders without crashing and matches snapshot 1`] = `
<Fragment>
  <PrimaryNav />
  <div
    className="l-container--sidebar"
  >
    <div
      className="l-side"
    >
      <SecondaryNav />
    </div>
    <div
      className="l-main"
    >
      <ErrorBoundary>
        <main
          className="main-content"
          id="main-content"
        />
      </ErrorBoundary>
    </div>
  </div>
</Fragment>
`;
